### PR TITLE
Check input size consistency in RNN and LSTM when using cuDNN

### LINKED ISSUE #5436: NStepLSTM silently fails with invalid in_size when cuDNN is enabled
## Problem description

When cuDNN is enabled NStepLSTM (and its variants), does not raise warning or error when `in_size` is incosistent with the actual input. It will run as if nothing is wrong, and then fails with a CUDA error. A random CUDA error seems to be raised (including `CUDNN_STATUS_EXECUTION_FAILED` and `CUDA_ERROR_ILLEGAL_ADDRESS`). 

## How to reproduce

1. Download and setup `text_classification` example

https://github.com/chainer/chainer/tree/v5.0.0rc1/examples/text_classification

2. Modify `in_size` of NStepLSTM to incosistent value.

```diff
-  self.encoder = L.NStepLSTM(n_layers, n_units, n_units, dropout)
+  self.encoder = L.NStepLSTM(n_layers, 50, n_units, dropout)
```

https://github.com/chainer/chainer/blob/v5.0.0rc1/examples/text_classification/nets.py#L129

3. Run `train_text_classifier.py`

```bash
python train_text_classifier.py --gpu 0 --model rnn --batchsize 256 --epoch 100
```

I set larger `batchsize` so that failure comes earlier.
It usually reproduces within first few epochs.

4. Confirmed that it does not reproduce without cuDNN

```bash
CHAINER_CUDNN=0 python train_text_classifier.py --gpu 0 --model rnn --batchsize 256 --epoch 100
```

It should (properly) raise error that input is incosistent.

I confirmed this behavior in Chainer v5.0.0rc1 and cupy v5.0.0rc1 with cuDNN 7.1.3.16-1+cuda9.1 and CUDA 9.1 on Ubuntu 16.04.

## Suggestion

I suggest we make NStepLSTM to do input size check or make documentation that recommend user to turn off cuDNN if weird CUDA error is raised, or possibly both.

It was super hard to figure out what is wrong and seriously need solving. (cuda-memcheck seems to point at random function as well)

### PR DESCRIPTION
**Pull request by takagi**
Close #5436.

This PR makes RNN and LSTM functions check input size consistency with their input values and weights when using cuDNN.

- `F.n_step_rnn`
- `F.n_step_birnn`
- `F.n_step_lstm`
- `F.n_step_bilstm`

### CODE REVIEW

**Comment by hvy** [comment]
Should we do this check for non-cuDNN paths as well (i.e. CPU or CUDA without cuDNN) or are the size constraints different in those cases?

**Comment by takagi** [comment]
In such case, `F.linear` used in these functions causes an error in its `check_type_forward` method, so we need no additional check.

**Comment by hvy** [comment]
Ah I see, thanks for the explanation. I think though that the error message that you introduce is a lot friendlier to the to user in these cases as well, since the linear functions is an implementation detail. I also think that the code is more readable with a unified shape validation before entering the device branch logic as well.

**Comment by takagi** [comment]
I got it, will fix the code!

**Comment by hvy** [comment]
jenkins, test this please.

File: chainer/functions/connection/n_step_lstm.py [hunk 1]
```diff
@@ -403,6 +403,13 @@ def n_step_lstm_base(
             'Use chainer.using_config')
         argument.assert_kwargs_empty(kwargs)
 
+    # Check input size consistency with xs and ws here.
+    x_in = xs[0].shape[1]
+    w_in = ws[0][0].shape[1]
+    if x_in != w_in:
+        raise ValueError('Inconsistent input size in input values and weight '
+                         'parameters')
```
Comment (hvy) [inline chainer/functions/connection/n_step_lstm.py]: How about including the actual values to make the message a bit more friendly?

File: chainer/functions/connection/n_step_rnn.py [hunk 2]
```diff
@@ -633,6 +633,13 @@ def n_step_rnn_base(n_layers, dropout_ratio, hx, ws, bs, xs,
         raise ValueError('Invalid activation: "%s". Please select from [%s]'
                          % (activation, candidate))
 
+    # Check input size consistency with xs and ws.
+    x_in = xs[0].shape[1]
+    w_in = ws[0][0].shape[1]
+    if x_in != w_in:
+        raise ValueError('Inconsistent input size in input values and weight '
+                         'parameters')
```
Comment (hvy) [inline chainer/functions/connection/n_step_rnn.py]: Ditto.

File: chainer/functions/connection/n_step_lstm.py (hunk 1 above)
Comment (takagi) [reply chainer/functions/connection/n_step_lstm.py]: Fixed!

File: chainer/functions/connection/n_step_rnn.py (hunk 2 above)
Comment (takagi) [reply chainer/functions/connection/n_step_rnn.py]: Fixed!

**Comment by hvy** [comment]
jenkins, test this please.

File: chainer/functions/connection/n_step_lstm.py (hunk 1 above)
Comment (hvy) [reply chainer/functions/connection/n_step_lstm.py]: Thanks for the quick fix!

**Comment by takagi** [comment]
I will fix the testing tolerance of `F.triplet` double backward as another issue.

**Comment by hvy** [comment]
LGTM!

**Pull request merged**